File: python/kth_largest_element_in_an_array.py
import heapq


class Solution:
    def findKthLargest(self, nums: list[int], k: int) -> int:
        """time complexity: O(klon(n)) where n is the length of nums, and k is the kth largest element"""
        min_heap: list[int] = nums
        heapq.heapify(min_heap)
        for _ in range(len(min_heap) - k):
            heapq.heappop(min_heap)
        return min_heap[0]


def test_kClosest_case_1():
    # arrange
    nums: list[int] = [3, 2, 1, 5, 6, 4]
    k: int = 2
    expected: int = 5

    # act
    solution = Solution()
    actual = solution.findKthLargest(nums, k)

    # assert
    assert expected == actual


def test_kClosest_case_2():
    # arrange
    nums: list[int] = [3, 2, 3, 1, 2, 4, 5, 5, 6]
    k: int = 4
    expected: int = 4

    # act
    solution = Solution()
    actual = solution.findKthLargest(nums, k)

    # assert
    assert expected == actual


def test_kClosest_case_3():
    # arrange
    nums: list[int] = [5, 2, 4, 1, 3, 6, 0]
    k: int = 4
    expected: int = 3

    # act
    solution = Solution()
    actual = solution.findKthLargest(nums, k)

    # assert
    assert expected == actual

File: rust/src/code/kth_largest_element_in_an_array.rs
struct Solution {}

impl Solution {
    // NOTE: time complexity O(nlogn)
    pub fn find_kth_largest(nums: Vec<i32>, k: i32) -> i32 {
        use std::collections::BinaryHeap;

        let heap = BinaryHeap::from(nums);
        let v = heap
            .into_sorted_vec()
            .into_iter()
            .rev()
            .collect::<Vec<i32>>();

        v[k as usize - 1]
    }
}

#[cfg(test)]
mod tests {
    use super::Solution;

    #[test]
    fn find_kth_largest_case_1() {
        // arrange
        let nums: Vec<i32> = vec![3, 2, 1, 5, 6, 4];
        let k: i32 = 2;
        let expected: i32 = 5;

        // act
        let actual = Solution::find_kth_largest(nums, k);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn find_kth_largest_case_2() {
        // arrange
        let nums: Vec<i32> = vec![3, 2, 3, 1, 2, 4, 5, 5, 6];
        let k: i32 = 4;
        let expected: i32 = 4;

        // act
        let actual = Solution::find_kth_largest(nums, k);

        // assert
        assert_eq!(expected, actual);
    }
}
